src/debug_sdfs_ed64.c

```diff
@@ -0,0 +1,590 @@
+	// format so we need to interleave the crc to take a bit from each CRC per
+	// line
+	t = (uint64_t)dat_crc[3] << 32 | dat_crc[2];
+	y = (uint64_t)dat_crc[1] << 32 | dat_crc[0];
+
+	t = everdrive_sd_interleave_bits(t, y);;
+
+	y = t & 0x00000000FFFFFFFF;
+	t = t >> 32;
+	
+	t = everdrive_sd_interleave_bits(t, y);;
+
+	crc_out[0] = t >> 48;
+	crc_out[1] = t >> 32;
+	crc_out[2] = t >> 16;
+	crc_out[3] = t;
+}
+
+static uint8_t everdrive_sd_stop_transmission() {
+	uint8_t result;
+	uint16_t timeout = -1;
+
+	if ((everdrive_sd_active_mode & ED64_SD_MODE_ACCESS) == ED64_SD_MODE_NONE) return 0;
+
+	result = everdrive_sd_execute_command(ED64_SD_CMD12, 0);
+	if (result & ED64_SD_TIMEOUT) return result;
+
+	// Wait until the buffer is cleaned
+	set_everdrive_sd_bitlen(8);
+	while (everdrive_sd_read_data() != 0xff) {
+		if (!timeout--) {
+			debugf("Buffer timed out\n");
+			return 0xFF;
+		}
+	}
+
+	return 0;
+}
+
+static LBA_t everdrive_sd_address = 0;
+static uint8_t everdrive_sd_change_mode(uint32_t mode, LBA_t addr) {
+	uint8_t result;
+	if (
+		(everdrive_sd_active_mode & ED64_SD_MODE_ACCESS) == mode &&
+		everdrive_sd_address == addr
+	) {
+		return 0;
+	}
+
+	result = everdrive_sd_stop_transmission();
+	if (result & ED64_SD_TIMEOUT) return result;
+
+	result = everdrive_sd_execute_command(
+		mode == ED64_SD_MODE_BLOCK_WRITE ? ED64_SD_CMD25 : ED64_SD_CMD18,
+		(everdrive_sd_active_mode & ED64_SD_MODE_IS_HC) ? addr : (addr * 512)
+	);
+	if (result & ED64_SD_TIMEOUT) return result;
+
+	everdrive_sd_active_mode &= ~ED64_SD_MODE_ACCESS;
+	everdrive_sd_active_mode |= mode;
+	everdrive_sd_address = addr;
+	return 0;
+}
+
+// Everdrive OS already does this but this is still necessary to find out if
+// the card is HC or not. Might be a simpler way in practice to just read OCR
+// CMD58 does not seem to work on its own.
+static DSTATUS fat_disk_initialize_everdrive(void) {
+	uint32_t sd_rca;
+	uint8_t result;
+
+	// Initialize active mode
+	everdrive_sd_active_mode = ED64_SD_MODE_NONE | ED64_SD_MODE_CMD_READ;
+	everdrive_sd_set_mode(ED64_SD_MODE_CMD_WRITE);
+
+	// Set lo speed for initialization
+	everdrive_sd_config &= ~ED64_SD_CFG_SPEED;
+	io_write(ED64_BASE_ADDRESS + ED64_REGISTER_SD_STATUS, everdrive_sd_config);
+
+	// Put in idle
+	everdrive_sd_execute_command(ED64_SD_CMD0, 0);
+
+	// IF cond
+	result = everdrive_sd_execute_command(ED64_SD_CMD8, 0x1AA);
+
+	if(result & ED64_SD_TIMEOUT) return RES_ERROR;
+
+	if (everdrive_sd_resp_buff[3] != 1) {
+		debugf("SD card - voltage mismatch\n");
+		return RES_ERROR;
+	}
+
+	if(everdrive_sd_resp_buff[4] != 0xAA) {
+		debugf("SD card did not echo\n");
+		return RES_ERROR;
+	};
+
+	int num_retries = ED64_SD_ACMD41_TOUT_MS / ED64_SD_ACMD41_WAIT_MS;
+	while(1) {
+		if(!num_retries--) {
+			debugf("SD card did not respond\n");
+			return RES_ERROR;
+			break;
+		}
+
+		// Query with HCS and 3.2-3.4V
+		result = everdrive_sd_send_app_command(ED64_SD_CMD41, 0, 0x40300000);
```

Comment:
For instante, here you're not testing result for timeout, so in the next line you might be testing the response buffer from a previous command.

Reply:
Thanks! This was an oversight by me.